README.md
[![Sensu Bonsai Asset](https://img.shields.io/badge/Bonsai-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/sensu/sensu-aggregate-check)
# Sensu Go Aggregate Check Plugin

- [Overview](#overview)
- [Files](#files)
- [Usage examples](#usage-examples)
- [Configuration](#configuration)
  - [Sensu Go](#sensu-go)
    - [Asset registration](#asset-registration)
    - [Check definition](#check-definition)
    - [RBAC](#rbac)
  - [Sensu Core](#sensu-core)
- [Installation from source](#installation-from-source)
- [Additional notes](#additional-notes)
- [Contributing](#contributing)

### Overview

An aggregate makes it possible to treat the result of multiple disparate check executions executed across multiple disparate systems as a single result (Event). Aggregates are extremely useful in dynamic environments and/or environments that have a reasonable tolerance for failure. Aggregates should be used when a service can be considered healthy as long as a minimum threshold is satisfied (e.g. are at least 5 healthy web servers? are at least 70% of N processes healthy?).

This plugin allows you to query the Sensu Go Backend API for Events matching certain criteria (labels). This plugin generates a set of counters (e.g. events total, events in an OK state, etc) from the Events query and provides several CLI arguments to evaluate the computed aggregate counters (e.g. --warn-percent=75).

### Files

N/A

## Usage examples

### Help

```
The Sensu Go Event Aggregates Check plugin

Usage:
  sensu-aggregate-check [flags]

Flags:
  -H, --api-host string          Sensu Go Backend API Host (e.g. 'sensu-backend.example.com') (default "127.0.0.1")
  -k, --api-key string           Sensu Go Backend API Key
  -P, --api-pass string          Sensu Go Backend API Password (default "P@ssw0rd!")
  -p, --api-port string          Sensu Go Backend API Port (e.g. 4242) (default "8080")
  -u, --api-user string          Sensu Go Backend API User (default "admin")
  -l, --check-labels string      Sensu Go Event Check Labels to filter by (e.g. 'aggregate=foo')
  -C, --crit-count int           Critical threshold - count of Events in warning state
  -c, --crit-percent int         Critical threshold - % of Events in warning state
  -e, --entity-labels string     Sensu Go Event Entity Labels to filter by (e.g. 'aggregate=foo,app=bar')
  -h, --help                     help for sensu-aggregate-check
  -i, --insecure-skip-verify     skip TLS certificate verification (not recommended!)
  -n, --namespaces string        Comma-delimited list of Sensu Go Namespaces to query for Events (e.g. 'us-east-1,us-west-2') (default "default")
  -s, --secure                   Use TLS connection to API
  -S, --skip-silence             Skip silenced events (not recommended!)
  -t, --trusted-ca-file string   TLS CA certificate bundle in PEM format
  -W, --warn-count int           Warning threshold - count of Events in warning state
  -w, --warn-percent int         Warning threshold - % of Events in warning state
```

## Configuration

### Sensu Go
#### Asset registration

Assets are the best way to make use of this plugin. If you're not using an asset, please consider doing so! If you're using sensuctl 5.13 or later, you can use the following command to add the asset:

`sensuctl asset add sensu/sensu-aggregate-check`

If you're using an earlier version of sensuctl, you can download the asset definition from [this project's Bonsai asset index page][1] or one of the existing [releases][2] or create an executable script from this source.

#### Check definition

```yaml
api_version: core/v2
type: CheckConfig
metadata:
  namespace: default
  name: dummy-app-aggregate
spec:
  runtime_assets:
  - sensu-aggregate-check
  command: sensu-aggregate-check --api-user=foo --api-pass=bar --check-labels='aggregate=healthz,app=dummy' --warn-percent=75 --crit-percent=50
  subscriptions:
  - backend
  publish: true
  interval: 30
  handlers:
  - slack
  - pagerduty
  - email
```

### RBAC

It is advised to use [RBAC][3] to create a user scoped specifically for purposes such as this check and
to not re-use the admin account.  For this check, in particular, the account would need access to
list and retrieve events.  The example below shows how to create a limited-scope user and the necessary
role and role-binding resources to give it the required access.

```
$ sensuctl user create aggregate --password='4yva#ko!Yq'
Created

$ sensuctl role create get-events --verb list,get --resource events
Created

$ sensuctl role-binding create aggregate-get-events --role=get-events --user=aggregate
Created
```

Though you could use the user and password combination above with this check, the best practice
would be to use an [API key][4] instead.  You can create the API key with sensuctl:

```
$ sensuctl api-key grant aggregate
Created: /api/core/v2/apikeys/03f66dbf-6fe0-40d4-8174-95b5eab95649
```

The key (the text after [...]/apikeys/) above can be used with the `--api-key` argument in place of using `api-user` and `api-pass`.

### Sensu Core

N/A

## Installation from source and contributing

### Sensu Go

To build from source, from the local path of the sensu-aggregate-check repository:
```
go build
```

### Contributing

To contribute to this plugin, see [CONTRIBUTING](https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md)

[1]: https://bonsai.sensu.io/assets/sensu/sensu-aggregate-check
[2]: https://github.com/sensu/sensu-aggregate-check/releases
[3]: https://docs.sensu.io/sensu-go/latest/reference/rbac/
[4]: https://docs.sensu.io/sensu-go/latest/reference/apikeys/



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased
- Add --skip-silence option 

### Changed
- Updated README to include section on RBAC

## [0.1.4] - 2020-04-28

### Changed
- Fix slice out of bounds error

## [0.1.3] - 2020-04-21

### Changed
- Fixed GitHub release to use sha512 checksums
- Offer better error failure messages, necessary for debugging deployments

## [0.1.2] - 2020-04-13

### Changed
- Fixed GitHub release to build with Golang 1.14.x

## [0.1.1] - 2020-04-13

### Changed
- Fixed goreleaser

## [0.1.0] - 2020-04-13

### Added
- Add options to support secure API connections
- Add support for using API key in place of username/password

### Changed
- Move to the new Sensu SDK
- Move to Go modules
- Move from Travis to GitHub actions

## [0.0.7] - 2019-08-14

### Added

Open sourced this Asset.

## [0.0.6] - 2019-03-22

### Fixed
- Fixed readme markdown for Bonsai

## [0.0.5] - 2019-03-22

### Added
- Added an OK STDOUT message ("Everything is OK")

## [0.0.4] - 2019-03-22

### Fixed
- Fixed percent calculation

## [0.0.3] - 2019-03-22

### Fixed
- Fixed build tars, added a goreleaser config

## [0.0.2] - 2019-03-21

### Added
- Releasing to Bonsai

## [0.0.1] - 2019-03-21

### Added
- Initial release


main.go
package main

import (
	"bytes"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"strings"

	"github.com/sensu-community/sensu-plugin-sdk/sensu"
	corev2 "github.com/sensu/sensu-go/api/core/v2"
	"github.com/sensu/sensu-go/types"
)

// Config represents the check plugin config.
type Config struct {
	sensu.PluginConfig
	CheckLabels        string
	EntityLabels       string
	Namespaces         string
	APIHost            string
	APIPort            int
	APIUser            string
	APIPass            string
	APIKey             string
	Secure             bool
	SkipSilence        bool
	TrustedCAFile      string
	InsecureSkipVerify bool
	Protocol           string
	WarnPercent        int
	CritPercent        int
	WarnCount          int
	CritCount          int
}

// Auth represents the authentication info
type Auth struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	ExpiresAt    int64  `json:"expires_at"`
}

// Counters represents the analyzed components and statuses count
type Counters struct {
	Entities int
	Checks   int
	Ok       int
	Warning  int
	Critical int
	Unknown  int
	Silenced int
	Total    int
}

var (
	tlsConfig tls.Config

	plugin = Config{
		PluginConfig: sensu.PluginConfig{
			Name:     "sensu-aggregate-check",
			Short:    "The Sensu Go Event Aggregates Check plugin",
			Keyspace: "sensu.io/plugins/sensu-aggregate-check/config",
		},
	}

	options = []*sensu.PluginConfigOption{
		&sensu.PluginConfigOption{
			Path:      "check-labels",
			Env:       "",
			Argument:  "check-labels",
			Shorthand: "l",
			Default:   "",
			Usage:     "Sensu Go Event Check Labels to filter by (e.g. 'aggregate=foo')",
			Value:     &plugin.CheckLabels,
		},
		&sensu.PluginConfigOption{
			Path:      "entity-labels",
			Env:       "",
			Argument:  "entity-labels",
			Shorthand: "e",
			Default:   "",
			Usage:     "Sensu Go Event Entity Labels to filter by (e.g. 'aggregate=foo,app=bar')",
			Value:     &plugin.EntityLabels,
		},
		&sensu.PluginConfigOption{
			Path:      "namespaces",
			Env:       "",
			Argument:  "namespaces",
			Shorthand: "n",
			Default:   "default",
			Usage:     "Comma-delimited list of Sensu Go Namespaces to query for Events (e.g. 'us-east-1,us-west-2')",
			Value:     &plugin.Namespaces,
		},
		&sensu.PluginConfigOption{
			Path:      "api-host",
			Env:       "",
			Argument:  "api-host",
			Shorthand: "H",
			Default:   "127.0.0.1",
			Usage:     "Sensu Go Backend API Host (e.g. 'sensu-backend.example.com')",
			Value:     &plugin.APIHost,
		},
		&sensu.PluginConfigOption{
			Path:      "api-port",
			Env:       "",
			Argument:  "api-port",
			Shorthand: "p",
			Default:   8080,
			Usage:     "Sensu Go Backend API Port (e.g. 4242)",
			Value:     &plugin.APIPort,
		},
		&sensu.PluginConfigOption{
			Path:      "api-user",
			Env:       "SENSU_API_USER",
			Argument:  "api-user",
			Shorthand: "u",
			Default:   "admin",
			Usage:     "Sensu Go Backend API User",
			Value:     &plugin.APIUser,
		},
		&sensu.PluginConfigOption{
			Path:      "api-pass",
			Env:       "SENSU_API_PASSWORD",
			Argument:  "api-pass",
			Shorthand: "P",
			Default:   "P@ssw0rd!",
			Usage:     "Sensu Go Backend API Password",
			Value:     &plugin.APIPass,
		},
		&sensu.PluginConfigOption{
			Path:      "api-key",
			Env:       "SENSU_API_KEY",
			Argument:  "api-key",
			Shorthand: "k",
			Default:   "",
			Usage:     "Sensu Go Backend API Key",
			Value:     &plugin.APIKey,
		},
		&sensu.PluginConfigOption{
			Path:      "warn-percent",
			Env:       "",
			Argument:  "warn-percent",
			Shorthand: "w",
			Default:   0,
			Usage:     "Warning threshold - % of Events in warning state",
			Value:     &plugin.WarnPercent,
		},
		&sensu.PluginConfigOption{
			Path:      "crit-percent",
			Env:       "",
			Argument:  "crit-percent",
			Shorthand: "c",
			Default:   0,
			Usage:     "Critical threshold - % of Events in warning state",
			Value:     &plugin.CritPercent,
		},
		&sensu.PluginConfigOption{
			Path:      "warn-count",
			Env:       "",
			Argument:  "warn-count",
			Shorthand: "W",
			Default:   0,
			Usage:     "Warning threshold - count of Events in warning state",
			Value:     &plugin.WarnCount,
		},
		&sensu.PluginConfigOption{
			Path:      "crit-count",
			Env:       "",
			Argument:  "crit-count",
			Shorthand: "C",
			Default:   0,
			Usage:     "Critical threshold - count of Events in warning state",
			Value:     &plugin.CritCount,
		},
		&sensu.PluginConfigOption{
			Path:      "secure",
			Env:       "",
			Argument:  "secure",
			Shorthand: "s",
			Default:   false,
			Usage:     "Use TLS connection to API",
			Value:     &plugin.Secure,
		},
		&sensu.PluginConfigOption{
			Path:      "insecure-skip-verify",
			Env:       "",
			Argument:  "insecure-skip-verify",
			Shorthand: "i",
			Default:   false,
			Usage:     "skip TLS certificate verification (not recommended!)",
			Value:     &plugin.InsecureSkipVerify,
		},
		&sensu.PluginConfigOption{
			Path:      "trusted-ca-file",
			Env:       "",
			Argument:  "trusted-ca-file",
			Shorthand: "t",
			Default:   "",
			Usage:     "TLS CA certificate bundle in PEM format",
			Value:     &plugin.TrustedCAFile,
		},
		&sensu.PluginConfigOption{
			Path:      "skip-silence",
			Env:       "",
			Argument:  "skip-silence",
			Shorthand: "S",
			Default:   false,
			Usage:     "Skip silenced events (not recommended!)",
			Value:     &plugin.SkipSilence,
		},
	}
)

func main() {
	check := sensu.NewGoCheck(&plugin.PluginConfig, options, checkArgs, executeCheck, false)
	check.Execute()
}

func checkArgs(event *types.Event) (int, error) {
	if len(plugin.CheckLabels) == 0 {
		return sensu.CheckStateWarning, fmt.Errorf("--check-labels is required")
	}
	if plugin.Secure {
		plugin.Protocol = "https"
	} else {
		plugin.Protocol = "http"
	}
	if len(plugin.TrustedCAFile) > 0 {
		caCertPool, err := corev2.LoadCACerts(plugin.TrustedCAFile)
		if err != nil {
			return sensu.CheckStateWarning, fmt.Errorf("Error loading specified CA file")
		}
		tlsConfig.RootCAs = caCertPool
	}
	tlsConfig.InsecureSkipVerify = plugin.InsecureSkipVerify

	tlsConfig.BuildNameToCertificate()
	tlsConfig.CipherSuites = corev2.DefaultCipherSuites

	return sensu.CheckStateOK, nil
}

func authenticate() (Auth, error) {
	var auth Auth
	client := http.DefaultClient
	client.Transport = http.DefaultTransport

	if plugin.Secure {
		client.Transport.(*http.Transport).TLSClientConfig = &tlsConfig
	}

	req, err := http.NewRequest(
		"GET",
		fmt.Sprintf("%s://%s:%d/auth", plugin.Protocol, plugin.APIHost, plugin.APIPort),
		nil,
	)
	if err != nil {
		return auth, fmt.Errorf("error generating auth request: %v", err)
	}

	req.SetBasicAuth(plugin.APIUser, plugin.APIPass)

	resp, err := client.Do(req)
	if err != nil {
		return auth, fmt.Errorf("error executing auth request: %v", err)
	}

	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return auth, fmt.Errorf("error reading auth response: %v", err)
	}

	if strings.HasPrefix(string(body), "Unauthorized") {
		return auth, fmt.Errorf("authorization failed for user %s", plugin.APIUser)
	}

	err = json.NewDecoder(bytes.NewReader(body)).Decode(&auth)

	if err != nil {
		trim := 64
		return auth, fmt.Errorf("error decoding auth response: %v\nFirst %d bytes of response: %s", err, trim, trimBody(body, trim))
	}

	return auth, err
}

func parseLabelArg(labelArg string) map[string]string {
	labels := map[string]string{}

	pairs := strings.Split(labelArg, ",")

	for _, pair := range pairs {
		parts := strings.Split(pair, "=")
		if len(parts) == 2 {
			labels[parts[0]] = parts[1]
		}
	}

	return labels
}

func filterEvents(events []*types.Event) []*types.Event {
	result := []*types.Event{}

	cLabels := parseLabelArg(plugin.CheckLabels)
	eLabels := parseLabelArg(plugin.EntityLabels)

	for _, event := range events {
		selected := true

		for key, value := range cLabels {
			if event.Check.ObjectMeta.Labels[key] != value {
				selected = false
				break
			}
		}

		if selected {
			for key, value := range eLabels {
				if event.Entity.ObjectMeta.Labels[key] != value {
					selected = false
					break
				}
			}
		}

		if selected {
			result = append(result, event)
		}
	}

	return result
}

func getEvents(auth Auth, namespace string) ([]*types.Event, error) {
	client := http.DefaultClient
	client.Transport = http.DefaultTransport

	url := fmt.Sprintf("%s://%s:%d/api/core/v2/namespaces/%s/events", plugin.Protocol, plugin.APIHost, plugin.APIPort, namespace)
	events := []*types.Event{}

	if plugin.Secure {
		client.Transport.(*http.Transport).TLSClientConfig = &tlsConfig
	}

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return events, fmt.Errorf("error creating GET request for %s: %v", url, err)
	}

	if len(plugin.APIKey) == 0 {
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", auth.AccessToken))
	} else {
		req.Header.Set("Authorization", fmt.Sprintf("Key %s", plugin.APIKey))
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return events, fmt.Errorf("error executing GET request for %s: %v", url, err)
	}

	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return events, fmt.Errorf("error reading response body during getEvents: %v", err)
	}

	err = json.Unmarshal(body, &events)
	if err != nil {
		trim := 64
		return events, fmt.Errorf("error unmarshalling response during getEvents: %v\nFirst %d bytes of response: %s", err, trim, trimBody(body, trim))
	}

	result := filterEvents(events)

	return result, err
}

func executeCheck(event *types.Event) (int, error) {
	var autherr error
	auth := Auth{}

	if len(plugin.APIKey) == 0 {
		auth, autherr = authenticate()

		if autherr != nil {
			return sensu.CheckStateUnknown, autherr
		}
	}

	events := []*types.Event{}

	for _, namespace := range strings.Split(plugin.Namespaces, ",") {
		selected, err := getEvents(auth, namespace)

		if err != nil {
			return sensu.CheckStateUnknown, err
		}

		for _, event := range selected {
			events = append(events, event)
		}
	}

	counters := Counters{}

	entities := map[string]string{}
	checks := map[string]string{}

	for _, event := range events {
		entities[event.Entity.ObjectMeta.Name] = ""
		checks[event.Check.ObjectMeta.Name] = ""

		if len(event.Check.GetSilenced()) > 0 && plugin.SkipSilence {
			counters.Silenced++
			continue
		}

		switch event.Check.Status {
		case 0:
			counters.Ok++
		case 1:
			counters.Warning++
		case 2:
			counters.Critical++
		default:
			counters.Unknown++
		}

		counters.Total++
	}

	counters.Entities = len(entities)
	counters.Checks = len(checks)

	fmt.Printf("Counters: %+v\n", counters)

	if counters.Total == 0 {
		fmt.Printf("WARNING: No Events returned for Aggregate\n")
		return sensu.CheckStateWarning, nil
	}

	percent := int((float64(counters.Ok) / float64(counters.Total)) * 100)

	fmt.Printf("Percent OK: %v\n", percent)

	if plugin.CritPercent != 0 {
		if percent <= plugin.CritPercent {
			fmt.Printf("CRITICAL: Less than %d%% percent OK (%d%%)\n", plugin.CritPercent, percent)
			return sensu.CheckStateCritical, nil
		}
	}

	if plugin.WarnPercent != 0 {
		if percent <= plugin.WarnPercent {
			fmt.Printf("WARNING: Less than %d%% percent OK (%d%%)\n", plugin.WarnPercent, percent)
			return sensu.CheckStateWarning, nil
		}
	}

	if plugin.CritCount != 0 {
		if counters.Critical >= plugin.CritCount {
			fmt.Printf("CRITICAL: %d or more Events are in a Critical state (%d)\n", plugin.CritCount, counters.Critical)
			return sensu.CheckStateCritical, nil
		}
	}

	if plugin.WarnCount != 0 {
		if counters.Warning >= plugin.WarnCount {
			fmt.Printf("WARNING: %d or more Events are in a Warning state (%d)\n", plugin.WarnCount, counters.Warning)
			return sensu.CheckStateWarning, nil
		}
	}

	fmt.Printf("Everything is OK\n")

	return sensu.CheckStateOK, nil
}
func trimBody(body []byte, maxlen int) string {
	if len(string(body)) < maxlen {
		maxlen = len(string(body))
	}

	return string(body)[0:maxlen]
}
